hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -367,16 +366,35 @@ private String[] authorityParts(URI uri) throws InvalidUriAuthorityException, In
   public boolean getIsNamespaceEnabled(TracingContext tracingContext)
       throws AzureBlobFileSystemException {
     try {
-      return this.isNamespaceEnabled.toBoolean();
+      return getNamespaceConfig();
     } catch (TrileanConversionException e) {
       LOG.debug("isNamespaceEnabled is UNKNOWN; fall back and determine through"
           + " getAcl server call", e);
     }
 
-    isNamespaceEnabled = Trilean.getTrilean(NamespaceUtil.isNamespaceEnabled(client, tracingContext));
+    try {
+      LOG.debug("Get root ACL status");
```

Comment:
I think this should be synchronized, with the first things in the sync block checking for the isNamespaceEnabled flag to see if it is now valid.

Lots of fs operations do this check, and so you could have multiple threads all triggering duplicate HTTP requests.

also, mark the field as volatile.

Reply:
Really great catch! Thanks! Have taken it.